### ISSUE TITLE
FlashAttention-2 backprop question

### ISSUE THREAD

**Issue by DianCh** [Dao-AILab/flash-attention#1012]
Hi authors! I've noticed that unlike FlashAttention-1, in FlashAttention-2's backward pass the attention scores are computed from exponentials WIHTOUT subtracting the max term intended for numerical stability:

(here the max number is added back to the `L_i` cache)
<img width="317" alt="Screenshot 2024-06-26 at 10 27 02 PM" src="https://github.com/Dao-AILab/flash-attention/assets/32043460/2321b6bb-46c3-43a6-9268-c4aa4e86381f">

<img width="378" alt="Screenshot 2024-06-26 at 10 19 05 PM" src="https://github.com/Dao-AILab/flash-attention/assets/32043460/1e182702-ada1-4a45-a6bb-316d66f479b9">

May I ask why this change? And why not still store `l_i` and `m_i` like in FA-1, so that this backward pass can still have the adjusted range for numerical stability (I thought the extra vector doesn't take up much memory but maybe you have other reasons?)?

Thank you!!

**Comment by tridao** [comment]
We subtract the logsumexp (L_i), which is the log of the softmax denominator. That gives you the right answer immediately, instead of subtracting the max, exponentiate, then divide by the sum.
Also logsumexp >= max, so it's numerically stable (i.e. exponential won't overflow).

**Comment by DianCh** [comment]
Oh I see. We subtract the max for numerical stability of individual exp terms in forward pass, but now this P_i won't overflow with the `-logsumexp`, even though mathematically it doesn't have the max subtraction.

Thank you for such prompt reply!
